presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolAllocator.java

```diff
@@ -126,9 +126,9 @@ public Symbol newSymbol(Field field)
         return newSymbol(nameHint, field.getType());
     }
 
-    public Map<Symbol, Type> getTypes()
+    public TypeProvider getTypes()
     {
-        return symbols;
+        return new TypeProvider(symbols);
```

Comment:
I am concerned about the cost of repeated copying of the map. Previously, we were happily sharing same (mutable) map. In Stats code, we ended up with `PMap`s.

Reply:
Fair. I'll see if I can make it work with a shared mutable map. This will all go away when I add types to IR expressions anyway.

Reply:
It's a little tricky, though. Not every place that creates a TypeProvider starts from a SymbolAllocator.